### PR TITLE
datapath/linux/route: use upstream function to list rules

### PR DESCRIPTION
**Pull request by florianl**
Signed-off-by: Lehner Florian <dev@der-flo.net>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

This PR resolves an open TODO from the code base. The referenced PR got merged upstream, so there is no need to maintain these constants in this code base.

### CODE REVIEW

File: pkg/datapath/linux/route/route_linux.go [hunk 1]
```diff
@@ -407,25 +407,6 @@ func ListRules(family int, filter *Rule) ([]netlink.Rule, error) {
...
 	}
 
-	const (
-		// RT_FILTER_PRIORITY is a flag that can be specified to signal
-		// filtering against rules with 'Priority' specified. Note: this count
-		// starts from where netlink stops, see
-		// https://github.com/vishvananda/netlink/blob/d71301a47b607450337d920f260f3dc76481298e/route_linux.go#L25
-		//
-		// TODO: Remove this function when the upstream PR has been merged:
-		// https://github.com/vishvananda/netlink/pull/538
```
Comment (christarazi) [inline pkg/datapath/linux/route/route_linux.go:417]: Looks like this whole function can be replaced with the upstream version, as https://github.com/vishvananda/netlink/pull/538 was merged.

**Review by christarazi — CHANGES_REQUESTED** [review]
Thanks for the PR! This one dropped off my radar, but https://github.com/vishvananda/netlink/pull/538 has been merged, so we don't need to use the upstream constants anymore; we can simply replace the entire function. Would you mind doing that?

File: pkg/datapath/linux/route/route_linux.go (hunk 1 above)
Comment (florianl) [reply pkg/datapath/linux/route/route_linux.go:417]: Sorry - I didn't spot this in the first place. Updated the PR accordingly.

**Comment by pchaigno** [comment]
test-me-please

**Review by christarazi — APPROVED** [review]
LGTM! Could you update the PR title to reflect the actual changes? Should be good to go after that.

**Comment by pchaigno** [comment]
There's one failure in the privileged unit tests that seems legitimate:
```
	 START: route_linux_test.go:245: RouteSuitePrivileged.TestListRules
	 Inside new network ns
	 Closed new network ns
	 Set back to previous network ns
	 ... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x438DB4)
	 
	 /usr/local/go/src/runtime/panic.go:969
	   in gopanic
	 /usr/local/go/src/runtime/panic.go:212
	   in panicmem
	 /usr/local/go/src/runtime/signal_unix.go:720
	   in sigpanic
	 route_linux.go:402
	   in ListRules
	 route_linux_test.go:280
	   in runListRules
	 route_linux_test.go:254
	   in testListRules4
	 route_linux_test.go:246
	   in RouteSuitePrivileged.TestListRules
	 /usr/local/go/src/reflect/value.go:336
	   in Value.Call
	 /usr/local/go/src/runtime/asm_amd64.s:1374
	   in goexit
	 PANIC: route_linux_test.go:245: RouteSuitePrivileged.TestListRules
```
The net-next failures are known flake https://github.com/cilium/cilium/issues/13062.

**Comment by tklauser** [comment]
test-me-please

**Review by tklauser — APPROVED** [review]
Nice!

**Comment by tklauser** [comment]
The `RuntimePrivilegedUnitTests` failure looks legitimate:

https://jenkins.cilium.io/job/Cilium-PR-Runtime-4.9/2009/testReport/junit/(root)/Suite-runtime/RuntimePrivilegedUnitTests_Run_Tests/

<details>

```
 START: route_linux_test.go:245: RouteSuitePrivileged.TestListRules
	 Inside new network ns
	 route_linux_test.go:246:
	     testListRules4(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 1, got: 4
	   []netlink.Rule{
	 + 	s"ip rule -1: from <nil> table 255",
	   	{Priority: 1, Table: 1, Mark: -1, Mask: -1, ...},
	 + 	s"ip rule 32766: from <nil> table 254",
	 + 	s"ip rule 32767: from <nil> table 253",
	   }
	 
	 
	 
	 route_linux_test.go:246:
	     testListRules4(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 1, got: 4
	   []netlink.Rule{
	 + 	s"ip rule -1: from <nil> table 255",
	   	{Priority: 1, Table: 1, Mark: -1, Mask: -1, ...},
	 + 	s"ip rule 32766: from <nil> table 254",
	 + 	s"ip rule 32767: from <nil> table 253",
	   }
	 
	 
	 
	 route_linux_test.go:246:
	     testListRules4(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 2, got: 5
	   []netlink.Rule{
	 + 	s"ip rule -1: from <nil> table 255",
	   	{Priority: 1, Table: 1, Mark: -1, Mask: -1, ...},
	   	{Priority: 1, Table: 1, Mark: -1, Mask: -1, ...},
	 + 	s"ip rule 32766: from <nil> table 254",
	 + 	s"ip rule 32767: from <nil> table 253",
	   }
	 
	 
	 
	 route_linux_test.go:246:
	     testListRules4(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 1, got: 5
	   []netlink.Rule{
	 + 	s"ip rule -1: from <nil> table 255",
	 + 	s"ip rule 1: from <nil> table 1",
	   	{Priority: 1, Table: 1, Mark: -1, Mask: -1, ...},
	 + 	s"ip rule 32766: from <nil> table 254",
	 + 	s"ip rule 32767: from <nil> table 253",
	   }
	 
	 
	 
	 route_linux_test.go:246:
	     testListRules4(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 4, got: 7
	   []netlink.Rule{
	 + 	s"ip rule -1: from <nil> table 255",
	   	{Priority: 5, Table: 1, Mark: -1, Mask: -1, ...},
	   	{Priority: 5, Table: 2, Mark: -1, Mask: -1, ...},
	   	{Priority: 5, Table: 3, Mark: -1, Mask: -1, ...},
	   	{Priority: 5, Table: 4, Mark: -1, Mask: -1, ...},
	 + 	s"ip rule 32766: from <nil> table 254",
	 + 	s"ip rule 32767: from <nil> table 253",
	   }
	 
	 
	 
	 route_linux_test.go:246:
	     testListRules4(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 1, got: 4
	   []netlink.Rule{
	 + 	s"ip rule -1: from <nil> table 255",
	   	{Priority: 1, Table: 199, Mark: -1, Mask: -1, ...},
	 + 	s"ip rule 32766: from <nil> table 254",
	 + 	s"ip rule 32767: from <nil> table 253",
	   }
	 
	 
	 
	 route_linux_test.go:246:
	     testListRules4(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 1, got: 6
	   []netlink.Rule{
	   	{Priority: 1, Table: 1, Mark: -1, Mask: 5, ...},
	 + 	s"ip rule 9: from <nil> table 2004",
	 + 	s"ip rule 10: from <nil> table 2005",
	 + 	s"ip rule 100: from <nil> table 255",
	 + 	s"ip rule 32766: from <nil> table 254",
	 + 	s"ip rule 32767: from <nil> table 253",
	   }
	 
	 
	 
	 route_linux_test.go:246:
	     testListRules4(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 1, got: 6
	   []netlink.Rule{
	   	{Priority: 1, Table: 1, Mark: 187, Mask: 255, ...},
	 + 	s"ip rule 9: from <nil> table 2004",
	 + 	s"ip rule 10: from <nil> table 2005",
	 + 	s"ip rule 100: from <nil> table 255",
	 + 	s"ip rule 32766: from <nil> table 254",
	 + 	s"ip rule 32767: from <nil> table 253",
	   }
	 
	 
	 
	 Closed new network ns
	 Set back to previous network ns
	 Inside new network ns
	 route_linux_test.go:247:
	     testListRules6(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 1, got: 3
	   []netlink.Rule{
	 + 	s"ip rule -1: from <nil> table 255",
	   	{Priority: 1, Table: 1, Mark: -1, Mask: -1, ...},
	 + 	s"ip rule 32766: from <nil> table 254",
	   }
	 
	 
	 
	 route_linux_test.go:247:
	     testListRules6(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 1, got: 3
	   []netlink.Rule{
	 + 	s"ip rule -1: from <nil> table 255",
	   	{Priority: 1, Table: 1, Mark: -1, Mask: -1, ...},
	 + 	s"ip rule 32766: from <nil> table 254",
	   }
	 
	 
	 
	 route_linux_test.go:247:
	     testListRules6(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 2, got: 4
	   []netlink.Rule{
	 + 	s"ip rule -1: from <nil> table 255",
	   	{Priority: 1, Table: 1, Mark: -1, Mask: -1, ...},
	   	{Priority: 1, Table: 1, Mark: -1, Mask: -1, ...},
	 + 	s"ip rule 32766: from <nil> table 254",
	   }
	 
	 
	 
	 route_linux_test.go:247:
	     testListRules6(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 1, got: 4
	   []netlink.Rule{
	 + 	s"ip rule -1: from <nil> table 255",
	 + 	s"ip rule 1: from <nil> table 1",
	   	{Priority: 1, Table: 1, Mark: -1, Mask: -1, ...},
	 + 	s"ip rule 32766: from <nil> table 254",
	   }
	 
	 
	 
	 route_linux_test.go:247:
	     testListRules6(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 4, got: 6
	   []netlink.Rule{
	 + 	s"ip rule -1: from <nil> table 255",
	   	{Priority: 5, Table: 1, Mark: -1, Mask: -1, ...},
	   	{Priority: 5, Table: 2, Mark: -1, Mask: -1, ...},
	   	{Priority: 5, Table: 3, Mark: -1, Mask: -1, ...},
	   	{Priority: 5, Table: 4, Mark: -1, Mask: -1, ...},
	 + 	s"ip rule 32766: from <nil> table 254",
	   }
	 
	 
	 
	 route_linux_test.go:247:
	     testListRules6(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 1, got: 3
	   []netlink.Rule{
	 + 	s"ip rule -1: from <nil> table 255",
	   	{Priority: 1, Table: 199, Mark: -1, Mask: -1, ...},
	 + 	s"ip rule 32766: from <nil> table 254",
	   }
	 
	 
	 
	 route_linux_test.go:247:
	     testListRules6(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 1, got: 5
	   []netlink.Rule{
	   	{Priority: 1, Table: 1, Mark: -1, Mask: 5, ...},
	 + 	s"ip rule 9: from <nil> table 2004",
	 + 	s"ip rule 10: from <nil> table 2005",
	 + 	s"ip rule 100: from <nil> table 255",
	 + 	s"ip rule 32766: from <nil> table 254",
	   }
	 
	 
	 
	 route_linux_test.go:247:
	     testListRules6(c)
	 route_linux_test.go:498:
	     c.Errorf("expected len: %d, got: %d\n%s\n", len(wantRules), len(rules), diff)
	 ... Error: expected len: 1, got: 5
	   []netlink.Rule{
	   	{Priority: 1, Table: 1, Mark: 187, Mask: 255, ...},
	 + 	s"ip rule 9: from <nil> table 2004",
	 + 	s"ip rule 10: from <nil> table 2005",
	 + 	s"ip rule 100: from <nil> table 255",
	 + 	s"ip rule 32766: from <nil> table 254",
	   }
	 
	 
	 
	 Closed new network ns
	 Set back to previous network ns
	 FAIL: route_linux_test.go:245: RouteSuitePrivileged.TestListRules
	 
	 START: route_linux_test.go:68: RouteSuitePrivileged.TestReplaceNexthopRoute
	 PASS: route_linux_test.go:68: RouteSuitePrivileged.TestReplaceNexthopRoute	0.001s
	 
	 START: route_linux_test.go:123: RouteSuitePrivileged.TestReplaceRoute
	 PASS: route_linux_test.go:123: RouteSuitePrivileged.TestReplaceRoute	0.001s
	 
	 START: route_linux_test.go:173: RouteSuitePrivileged.TestReplaceRule
	 PASS: route_linux_test.go:173: RouteSuitePrivileged.TestReplaceRule	0.001s
	 
	 START: route_linux_test.go:182: RouteSuitePrivileged.TestReplaceRule6
	 PASS: route_linux_test.go:182: RouteSuitePrivileged.TestReplaceRule6	0.001s
	 
	 START: route_linux_test.go:191: RouteSuitePrivileged.TestRule_String
	 PASS: route_linux_test.go:191: RouteSuitePrivileged.TestRule_String	0.000s
	 
	 OOPS: 5 passed, 1 FAILED
	 --- FAIL: Test (0.04s)
	 FAIL
	 coverage: 3.5% of statements in ./...
	 FAIL	github.com/cilium/cilium/pkg/datapath/linux/route	0.085s
	 FAIL
	 Makefile:146: recipe for target 'tests-privileged' failed
	 make: Leaving directory '/home/vagrant/go/src/github.com/cilium/cilium'
	 
Stderr:
 	 make: *** [tests-privileged] Error 1
```
</details>

File: pkg/datapath/linux/route/route_linux_test.go [hunk 2]
```diff
@@ -279,6 +279,16 @@ func runListRules(c *C, family int, fakeIP, fakeIP2 *net.IPNet) {
...
 
+	ruleMinus1 := netlink.Rule{
+		Priority: -1, Mark: -1, Mask: -1, Goto: -1, Table: 255, Flow: -1, SuppressIfgroup: -1, SuppressPrefixlen: -1,
+	}
+	rule32766 := netlink.Rule{
+		Priority: 32766, Mark: -1, Mask: -1, Goto: -1, Table: 254, Flow: -1, SuppressIfgroup: -1, SuppressPrefixlen: -1,
+	}
+	rule32767 := netlink.Rule{
+		Priority: 32767, Mark: -1, Mask: -1, Goto: -1, Table: 253, Flow: -1, SuppressIfgroup: -1, SuppressPrefixlen: -1,
+	}
```
Comment (christarazi) [inline pkg/datapath/linux/route/route_linux_test.go]: Do we know why we need these now? The original function should have been equivalent to the upstream function that we're now using.

Comment (florianl) [reply pkg/datapath/linux/route/route_linux_test.go]: I did not set the mask for the filter correctly. That's why these additional rules were returned. Fix it now.

**Comment by pchaigno** [comment]
test-me-please

**Review by christarazi — APPROVED** [review]
Ah I see, it's a bit subtle in the end. Nice catch and thanks for digging in on this!

**Pull request merged**